openedx/core/djangoapps/content_libraries/tasks.py

```diff
@@ -39,3 +78,281 @@ def on_progress(block_key, block_num, block_count, exception=None):
+                    # Make sure the URL is one that will work from the user's browser when using the docker devstack
+                    field_value = blockstore_api.force_browser_url(field_value)
+            setattr(new_block, field_name, field_value)
+    new_block.save()
+    store.update_item(new_block, user_id)
+
+    if new_block.has_children:
+        # Delete existing children in the new block, which can be reimported again if they still exist in the
+        # source library
+        for existing_child_key in new_block.children:
+            store.delete_item(existing_child_key, user_id)
+        # Now import the children
+        for child in source_block.get_children():
+            _import_block(store, user_id, child, new_block_key)
+
+    return new_block_key
+
+
+def _filter_child(store, usage_key, capa_type):
+    """
+    Return whether this block is both a problem and has a `capa_type` which is included in the filter.
+    """
+    if usage_key.block_type != "problem":
+        return False
+
+    descriptor = store.get_item(usage_key, depth=0)
+    assert isinstance(descriptor, ProblemBlock)
+    return capa_type in descriptor.problem_types
+
+
+def _get_library(store, library_key):
+    """
+    Helper method to get either V1 or V2 library.
+
+    Given a library key like "library-v1:ProblemX+PR0B" (V1) or "lib:RG:rg-1" (v2), return the 'library'.
+
+    is_v2_lib (bool) indicates which library storage should be requested:
+        True - blockstore (V2 library);
+        False - modulestore (V1 library).
+
+    Returns None on error.
+    """
+    if isinstance(library_key, LibraryLocatorV2):
+        try:
+            return library_api.get_library(library_key)
+        except ContentLibrary.DoesNotExist:
+            return None
+    else:
+        try:
+            return store.get_library(
+                library_key, remove_version=False, remove_branch=False, head_validation=False
+            )
+        except ItemNotFoundError:
+            return None
+
+
+def _problem_type_filter(store, library, capa_type):
+    """ Filters library children by capa type."""
+    try:
+        search_engine = SearchEngine.get_search_engine(index="library_index")
+    except:  # pylint: disable=bare-except
+        search_engine = None
+    if search_engine:
+        filter_clause = {
+            "library": str(normalize_key_for_search(library.location.library_key)),
+            "content_type": ProblemBlock.INDEX_CONTENT_TYPE,
+            "problem_types": capa_type
+        }
+        search_result = search_engine.search(field_dictionary=filter_clause)
+        results = search_result.get('results', [])
+        return [LibraryUsageLocator.from_string(item['data']['id']) for item in results]
+    else:
+        return [key for key in library.children if _filter_child(store, key, capa_type)]
+
+
+def import_from_blockstore(user_id, store, dest_block, blockstore_block_ids):
```

Comment:
Note: we have this `import_from_blockstore` code both here and in library tools. Currently thinking about which place is the correct place to put this code, but just don't want to have redundant or unused code so it is worthwhile to remove one.

i'll put a comment where this code is duplicated.

Reply:
Good catch @connorhaugh .

My thinking is that `tasks.py` is the right place for this code rather than the library_tools service class. That's because `import_from_blockstore` isn't really a "service" that we want to provide to XBlocks. The service we want to provide is `update_children`, and `import_from_blockstore` is just an implementation detail to achieve that.

Also, from a practical standpoint, keeping this function in `library_tools.py` would prevent an import cycle: `library_tools:LibraryTools.update_children` -> `tasks.update_children_task` -> `library_tools:LibraryToolsService:import_from_blockstore`

Reply:
Done. Latest commit should fix tests.

FYI, two other changes I made:
* `update_children` is now `trigger_update_children_task`.
* there was a test that specifically tested `_import_from_blockstore` (which is now a private function in `tasks.py`), so I updated it to test the same functionality but through `trigger_update_children_task`.